### ISSUE TITLE
Electron Crashes in v7 and v8 with Yarn PnP enabled

### ISSUE THREAD

**Issue by Mike-Dax** [electron/electron#22472]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

I found an issue regarding this but it was closed incorrectly.

### Issue Details

* **Electron Version:**
  * 8.0.2
  * 7.1.14

* **Operating System:**
  * macOS 10.14.6
* **Last Known Working Electron version:**
  * 9.0.0-beta-1 through 9.0.0-beta-4 work

### Expected Behavior

Electron should boot and not crash immediately.

### Actual Behavior

Electron immediately crashes in the render processes.

### To Reproduce

I have a minimum reproduction repo here. Since yarn v2 is the 'change' here, it needs to be used to boot. If you have yarn 1.21.1, it should automatically use the vendored yarn provided.

```sh
git clone https://github.com/Mike-Dax/electron-yarn-pnp-crash.git
yarn install
yarn dev
```

### Screenshots

![image](https://user-images.githubusercontent.com/13504878/75648470-18463b00-5ca0-11ea-8b79-cfa6cb00075b.png)

### Additional Information

 Issue #22053 was correct in this being an issue.

https://github.com/electron/electron-quick-start/issues/359 should be closed, this isn't a template issue.

This seems to have been fixed in the Electron v9 beta, it would be nice for a backport to occur to v8 if that's at all possible.

Thank you!

**Comment by merceyz** [comment]
Output from `--enable-logging`

```
 [772:1227/224809.010:INFO:CONSOLE(11009)] "failed to compile wasm module: RangeError: WebAssembly.Compile is disallowed on the main thread, if the buffer size is larger than 4KB
. Use WebAssembly.compile, or compile on a worker thread.", source: C:\app\.pnp.js (11009)
[772:1227/224809.020:INFO:CONSOLE(11009)] "Assertion failed: binaryen setup failed (no wasm support?)", source: C:\app\.pnp.js (11009)
[772:1227/224809.020:INFO:CONSOLE(11009)] "Assertion failed: binaryen setup failed (no wasm support?)", source: C:\app\.pnp.js (11009)
[772:1227/224809.020:INFO:CONSOLE(11009)] "Uncaught abort("Assertion failed: binaryen setup failed (no wasm support?)") at Error
    at jsStackTrace (C:\app\.pnp.js:11009:13760)
    at stackTrace (C:\app\.pnp.js:11009:13931)
    at abort (C:\app\.pnp.js:11009:312117)
    at assert (C:\app\.pnp.js:11009:6807)
    at Object.Module.asm (C:\app\.pnp.js:11009:193504)
    at Object.<anonymous> (C:\app\.pnp.js:11009:280039)
    at __webpack_require__ (C:\app\.pnp.js:10937:30)
    at Module.<anonymous> (C:\app\.pnp.js:12440:14)
    at __webpack_require__ (C:\app\.pnp.js:10937:30)
    at C:\app\.pnp.js:11001:18", source: C:\app\.pnp.js (11009)
```

**Comment by Mike-Dax** [comment]
Interesting, it seems to be this:

https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-libzip/sources/libzipSync.js

Did synchronous compilation of WebAssembly modules greater than 4KB become allowed in Electron v9?

**Comment by MarshallOfSound** [comment]
So I'm going to float this out there and then close this issue a bit later.

Electron can't put resources towards fixing or maintaining compatibility with yarn pnp.  It likely will never work properly on end user machines (in a packaged state) and does things with module resolution and linking that we as a runtime can't start supporting.  Electron delibrately injects itself in the node module loader and has to wrap all calls to the filesystem, without this injection and wrapping Electron apps will not work.

TLDR: What pnp wants to do is fundamentally incompatible with some parts of the Electron app lifecycle and we can't put resources towards fixing it.

**Comment by Mike-Dax** [comment]
I assume the majority of users bundle their applications before deployment into Electron, they'd be able to bundle their app then run Electron without any PnP integration. Any difficulties regarding that should have issues opened in the relevant bundler plugins.

A flag on the Yarn side to disable `NODE_OPTIONS --require .pnp.js` injection when running binaries should make that easier to deal with.

This crash however seems specifically related to synchronous WebAssembly compilation, I can make a new issue regarding that, I'll investigate and produce a reproducible example without PnP.

**Comment by Mike-Dax** [comment]
Following up on this - we ended up running Yarn PnP on the webpack-dev-server side of things, providing that bundle to Electron during development.

We just had to explicitly delete the `NODE_OPTIONS` environment variable when running Electron. Electron as a result has no understanding or requirement to understand PnP.

In production we Webpack our sources and run it as per normal.

This gives us the best of all worlds!

**Comment by alecmev** [comment]
> We just had to explicitly delete the NODE_OPTIONS environment variable when running Electron. Electron as a result has no understanding or requirement to understand PnP.

For future readers, to be more specific, it's sufficient to add the following to `main`:

```js
process.env.NODE_OPTIONS = undefined;
```

Alternatively, run electron directly:

```sh
.yarn/unplugged/electron-npm-version-hash/node_modules/electron/dist/electron dist
```

**Comment by pomSense** [comment]
> > We just had to explicitly delete the NODE_OPTIONS environment variable when running Electron. Electron as a result has no understanding or requirement to understand PnP.
> 
> For future readers, to be more specific, it's sufficient to add the following to `main`:
> 
> ```js
> process.env.NODE_OPTIONS = undefined;
> ```
> 
> Alternatively, run electron directly:
> 
> ```shell
> .yarn/unplugged/electron-npm-version-hash/node_modules/electron/dist/electron dist
> ```

Thanks @alecmev this worked. For others, just double check the exact path as it may have changed. Mine was:
![image](https://user-images.githubusercontent.com/35952778/200053931-eee43bb5-e7db-48ae-97de-0302dbf86504.png)

And since I am using workspaces, I had to do the following after the server started:
```
../../../.yarn/unplugged/electron-npm-21.2.2-9d373f990b/node_modules/electron/dist/Electron.app/Contents/MacOS/Electron .
```
